### ISSUE TITLE
Agent destruction does not always result in termination event in Python

### ISSUE THREAD

**Issue by ykeuter** [Unity-Technologies/ml-agents#4864]
The destruction of an `Agent` by `Destroy(GameObject)` normally leads to a termination event in Python. The exception seems to be if the agent is the only one active in the scene: in this case there is no more 'step' info passed to Python, hence no termination reported. This could be by design, following the documentation:
[Destroying an Agent](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Learning-Environment-Design-Agents.md#destroying-an-agent)
This also means though that it is not possible to reset the environment through the Python API in case all agents are destroyed, since there is no way of knowing.

Could you confirm this is really by design? It seems like a common scenario to reset the environment from Python in case all agents are destroyed.

I use the package versions of ML-Agents Release 10.

**Comment by vincentpierre** [comment]
Hi, The short answer is that it is by design. When an Agent is terminated (in this case destroyed) it will send a message to a buffer to be sent later. The buffer is then sent to Python as soon as another Agent (Or the same Agent after respawn) needs to request a decision. This is to avoid sending messages that only contain information about the death of Agents and do not really need an input from Python.

In RL, the environment can be reset at any time by Python and in ML-Agents it is not necessary to manually reset the environment, it is assumed that when the game reached a terminal state (like there are no Agents in the scene for example) the game will reset itself and spawn new Agents.

In short, there should be no moment where the game no longer has any Agents taking decisions and is waiting on Python to reset. It is not assumed that Python will send a reset signal when no Agents are alive in the scene.

I hope this helps. What is your use case for this ?

**Comment by ykeuter** [comment]
Hi @vincentpierre - Thanks for your quick reply!

The use case I had, was that I wanted to evaluate a single agent in multiple training areas in parallel. Once all these areas had finished, Python would reset the environment to evaluate the next agent in the population.

Your design decision is clear though, so I will close this issue.

Regarding your explanation, I do see `get_steps` return values where `decision_steps` is empty and `terminal_steps` is not, which seems to contradict part of your statement. How can this be?
